src/hashing.rs

```diff
@@ -0,0 +1,495 @@
+//! is not sufficient), `D_1` and `D_2` behave as independent high-quality cryptographic hash functions, and generally
+//! retain the useful properties of `D`.
+//!
+//! [hmac]: https://en.wikipedia.org/wiki/HMAC#Design_principles "HMAC: Design principles"
+
+use std::{marker::PhantomData, ops::Deref};
+
+use blake2::VarBlake2b;
+use digest::Digest;
+use sha3::Sha3_256;
+use tari_utilities::ByteArray;
+
+use crate::{common::Blake256, errors::HashingError, keys::SecretKey};
+
+/// The `DomainSeparation` trait is used to inject domain separation tags into the [`DomainSeparatedHasher`] in a way
+/// that can be applied consistently, but without hard-coding anything into the hasher itself.
+///
+/// Using a trait is more flexible than const strings, and lets us leverage the type system to have more fine-grained
+/// control over allowable use cases.
+///
+/// For example, not all digest functions are suitable for use with the MAC generator provided in this crate. We can
+/// indicate this at _compile time_ by adding a trait bound that prevents a client using these functions. See
+/// [`MacDomain`] for details.
+pub trait DomainSeparation {
+    /// Returns the version number for the metadata tag
+    fn version() -> u8;
+    /// Returns the category label for the metadata tag. For example, `tari_hmac`
+    fn domain() -> &'static str;
+    /// The domain separation tag is defined as `{domain}.v{version}.{label}`, where the version and tag are
+    /// typically hard-coded into the implementing type, and the label is provided per specific application of the
+    /// domain
+    fn domain_separation_tag<S: AsRef<str>>(label: S) -> String {
+        format!("{}.v{}.{}", Self::domain(), Self::version(), label.as_ref())
+    }
+}
+
+//--------------------------------------     Domain Separated Hash   ---------------------------------------------------
+
+/// A hash value, guaranteed, as far as possible, to have been created using a hash function that has been randomly and
+/// independently sampled from an idealized set of hash functions.
+///
+/// This is modelled via the strategy of applying a
+/// domain separation tag that is unique for this hashing application (assuming clients make proper use of a unique
+/// label for every discrete hashing use-case in their applications).
+///
+/// `DomainSeparatedHash` implements `AsRef<u8>`, so it is easy to use this type as a slice, or you can discard the
+/// domain tag by calling [`DomainSeparatedHash::into_vec`].
+///
+/// The domain separation information is retained with the hash, and can be queried using
+/// [`DomainSeparatedHash::domain_separation_tag`].
+///
+/// To preserve the guarantee that the hash is properly domain separated, you cannot create an instance of this struct
+/// directly. It is the result of using [`DomainSeparatedHasher`].
+///
+/// For details and examples, see [`DomainSeparatedHasher`].
+pub struct DomainSeparatedHash {
+    hash: Vec<u8>,
+    tag: String,
```

Comment:
Both of these fields are heap-allocated. Since hashing is the bread and butter of crypto software, ideally these allocations should be avoided. `digest` provides the `Output<D> where D: Digest` type (alias to `generic-array` crate) that is stack-based and avoids the heap. 

As for the `tag`, perhaps it can be removed entirely as it probably won't ever be needed after hashing (if that info is really needed it could be preserved some other way e.g. a wrapper/new-type).

Reply:
https://github.com/tari-project/tari-crypto/issues/113